judoscale/celery/__init__.py

```diff
@@ -0,0 +1,32 @@
+import time
+
+from celery import Celery
+from celery.signals import before_task_publish
+
+from judoscale.celery.collector import CeleryMetricsCollector
+from judoscale.core.config import config as judoconfig
+from judoscale.core.reporter import reporter
+
+before_publish_handler = None
+
+
+def before_publish(collector: CeleryMetricsCollector):
+    def _before_publish(*args, routing_key, properties={}, **kwargs):
+        collector.queues.add(routing_key)
+        properties["published_at"] = time.time()
+
+    return _before_publish
+
+
+def judoscale_celery(celery: Celery, extra_config: dict = {}) -> None:
+    global before_publish_handler
+
+    celery.conf.task_send_sent_event = True
+
+    judoconfig.merge(extra_config)
+    collector = CeleryMetricsCollector(config=judoconfig, broker=celery)
+
+    before_publish_handler = before_task_publish.connect(before_publish(collector))
```

Comment:
Can you help me understand the difference between "before publish" and "task sent", and why we're handling both?

Reply:
Whoops — I forgot to refactor this bit of code.

`before_task_publish` is the signal we hook into to add the `published_at` timestamp to the task's properties. This is signal only exists within the process.

`task-sent` is received by any process that sets up a `app.events.Receiver` to explicitly receive that event. It's more like pubsub than a signal sent and handled within the same process.